### DISCUSSION TITLE
Create a persistent palette from process

### DISCUSSION THREAD

**Discussion by c-langlet** [brainvisa/brainvisa.github.io#77] [Q&A] [answered]
Hi,

How can i create a palette from a Brainvisa process, store it on disk and retrieve it from another process ? 
Currently I have a set of colors that I assign to a palette :  

    palette = a.createPalette('colormap')
    palette.setColors(colors=RGB_colors, color_mode='RGB')

Then I store it in a Volume (not sure if it's the best thing to do, maybe its better to create directly the Volume) and write it on disk as a png : 

    vol = aims.Volume(np.asarray(palette))
    aims.write(vol, '/my/path/colormap.png')

I'm not sure at this point if I can store my palette anywhere or if I should store it in ./anatomist  

Then, in another process, I want to be able to refer to this palette. Currently I try using :  

    myTexture.setPalette('colormap')

I also tried with getPalette, using the name of the palette ('colomap') or the full path of the volume ('/my/path/colormap.png'). I understand that anatomist may not know the palette 'colormap' because it was created in another instance of Anatomist and may not be a persistent palette, but how come I cannot retrieve the palette from my volume ? Should I use a.loadObject('/my/path/colormap.png') or aims.read('/my/path/colormap.png') (I can't remember at this point if I tried these solutions already) ? 

Sorry for some eventual typos, I wrote this without my source code. 

Thanks for the reply,
Clément.

**Comment by denisri** [comment, marked as answer]
Palettes created within an Anatomist session are not persistent, so if they have to be reused later, you have to save them and re-create them when needed. You almost have done it actually, I think. However what you have done with:
```
myTexture.setPalette('colormap')
```
should work within the same anatomist / brainvisa session instance, but not in another one.
There are 2 choices:
1. save the palette image in `$HOME/.anatomist/rgb/` (if the directory doesn't exist, you can create it). Palettes saved here will be loaded by anatomist at startup, or when using the "reload palettes" option (in the GUI, not sure if there is an API function for that in python, I don't remember but we can check that). The name of the palette is the palette image name (basename, without extension). This will work for you, but such a palette is a "personal" one, other users will not have it. Moreover if several process happen to create palettes this way with the same name but different contents, the last one will erase the older ones.
2. save the palette image at a reusable location (could be in brainvisa database, or whatever), and each time you need it in anatomist, check if it is here or reload it:
```
if a.getPalette('colormap') is None:
    palette = a.createPalette('colormap')
    # load RGB[A] colors...
    palette.setColors(colors=RGB_colors, color_mode='RGB')
myTexture.setPalette('colormap')
```
There was a way somewhere to load a palette data from a RGBA file but... I don't find it right now. I'll look for it...

**Reply by c-langlet** [reply]
I tried solution 1. and this worked only after using the "reload palettes" in the GUI, but not at Anatomist instance start up... So I guess if there is no API function to do that it's a dead end. 

Anyway, I'll go for the second solution, but I am interested in the name of that API function that can load a palette from a RGBA if you can find it.

**Comment by denisri** [comment]
1. It's not normal that personal palettes are not loaded at anatomist startup. I mean, "real" startup, not a reuse of the Anatomist instance singleton within the same Brainvisa session - but even there the palettes created previously should still be here... 
    Anyway you can force the reload using:
```
a.palettes().load(os.path.expanduser('~/.anatomist/rgb'))
```
2. You can do it after updating aims-free (pyaims) from the current master branch (*):
```
palette = a.createPalette('colormap')
aims.read('/my/path/colormap.png', object=palette, dtype='Volume_RGBA')
```

(*) the update is needed to fix a problem in `pyaims`: here we need to load an object of type `APalette`, which has no "direct" reader (in C++, it would be `Reader<APalette>` which isn't implemented), however `APalette` actually inherits `Volume_RGBA` (`Volume<AimsRGBA>` in C++), thus we can use a reader for the type `Volume_RGBA`. But the aims reader doesn't resolve inheritances in readers (C++ templates also don't) and we need to force the reader type to be used. Before the fix, the `object` type had priority over the passed `dtype`, which forbid to do what we want. Now `dtype` has priority.
